flang/lib/Evaluate/intrinsics.cpp

```diff
@@ -333,11 +333,15 @@ static const IntrinsicInterface genericIntrinsicFunction[]{
                 common::Intent::In, {ArgFlag::canBeNull}}},
         DefaultLogical, Rank::elemental, IntrinsicClass::inquiryFunction},
     {"atan", {{"x", SameFloating}}, SameFloating},
-    {"atand", {{"x", SameFloating}}, SameFloating},
     {"atan", {{"y", OperandReal}, {"x", OperandReal}}, OperandReal},
-    {"atand", {{"y", OperandReal}, {"x", OperandReal}}, OperandReal},
+    {"atand",
+        {{"y", SameReal, Rank::scalar, Optionality::required,
+             common::Intent::In},
+            {"x", SameReal, Rank::scalar, Optionality::optional,
+                common::Intent::In}},
+        SameReal, Rank::scalar, IntrinsicClass::inquiryFunction},
```

Comment:
For what where these semantics changes required?
This does not looks correct to me. There is a subtle by important difference between the standard writing `ATAND([Y], X)` and `ATAND(Y, X) or ATAND(X)`.

In the former, we would have to treat Y has having the OPTIONAL attribute, and need special handling when the actual argument is a pointer, allocatable or OPTIONAL dummy. See the little `(optional)` that the standard adds for such argument (e.g. STAT in ATOMIC_ADD).

In the later, there is no such semantics, and if Y is syntactically here, it cannot be a disassociated pointer or absent optional.

Thus, I think the new change is not correct.

I would also not update things from  `OperandReal` to `SameReal`, flang is relaxed with regard to the same kind rule and should upcast to the highest precision (because that is what some other compilers do with numerical intrinsics and may be expected). Was there a reason to differ from atan2 behavior here too?

Reply:
It is possible to have intrinsics with the same name but different para list here, like how it was done before. Having an `atand(X)`, and `atand(Y,X)`. But in the `intrinsicCall.cpp`, they seem only different from each other by `args.size()`. I do have another implementation by have both `atand(X)` and `atand(Y,X)`, and treat them both in `genAtand` by using `args.size()` for each case. I'm not sure if using `args.size()` is the safe/correct way to do this. I can use `atand([Y],X)` if `args.size()` could not be used.

The reason for me to use `atand(Y,[X])` here is that if X is not present, it would be treated as `atand(X)`. This definitely does not make sense in terms of naming, because I am actually using Y as X for the one input version `atand(X)`.

I will change it back from `SameReal` to `OperandReal` regard to the relaxation of same kind rule, and also update all other `atan` with the zero check.

Reply:
Using args.size() to distinguish here is fine and done in other cases like bessel_jn here https://github.com/llvm/llvm-project/blob/0f4e9e718b4e79d3f760fd7ea399c23fc97a7716/flang/lib/Optimizer/Builder/IntrinsicCall.cpp#L2200.